code.js

```diff
@@ -1,3 +1,41 @@
-function tsp_hk(distance_matrix) {
-    return -1;
+function tsp_hk(dm) {
+  let cities = dm.map((item, index) => index);
+  cities.sort();
+
+  if (cities.length === 1) return 0;
+
+  let cache = new Map();
+  let sumDist = Infinity;
+
+  for (let start of cities) {
+    sumDist = tsp(cities, start, dm, cache);
+    // let result = tsp(cities, start, dm, cache);
+    // sumDist = Math.min(sumDist, result);
+  }
+
+  return sumDist;
 }
+
+function tsp(cities, start, dm, cache) {
+  if (cities.length === 1) {
+    return dm[start][cities[0]];
+  } else {
+    
+    let key = JSON.stringify(cities) + start;
```

Comment:
What if `cities` is in a different order? Shouldn't that go to the same cache entry?